api/include/opentelemetry/metrics/meter_provider.h

```diff
@@ -20,15 +23,111 @@ class MeterProvider
 {
 public:
   virtual ~MeterProvider() = default;
+
+#if OPENTELEMETRY_ABI_VERSION_NO >= 2
+
+  /**
+   * Gets or creates a named Meter instance (ABI).
+   *
+   * @since ABI_VERSION 2
+   *
+   * @param[in] name Meter instrumentation scope
+   * @param[in] version Instrumentation scope version, optional
+   * @param[in] schema_url Instrumentation scope schema URL, optional
+   * @param[in] attributes Instrumentation scope attributes, optional
+   */
+  virtual nostd::shared_ptr<Meter> DoGetMeter(
```

Comment:
Is this intended to be public method?

Reply:
Still investigation options here:

* Naming this GetMeter() causes too much confusion with overloaded prototypes, so that {{"foo", 1}} does not use std::initializer but attempts to use this one instead, causing a break.
* If we keep GetMeter() only on the API, and have DoGetMeter() as virtual method, it can be made protected instead

Reply:
I am finding out that having too many overloaded prototypes, with optional parameters, present both at the API level and at the SDK level utterly confuses the compiler, when calling a method on the SDK class.

This is also confusing for humans, because nobody knows if GetMeter(prototype X) and GetMeter(prototype Y) are part of the API or the ABI.

To discuss, currently leaning towards:

* GetMeter() only on API, delegates to virtual DoGetMeter(ABI prototype)
* Only DoGetMeter() is present on the SDK class
* GetMeter() is public
* DoGetMeter() is protected

The point is to avoid a public GetMeter() method in the SDK that hides other GetMeter() from the API.

Given the complexity of:
* https://en.cppreference.com/w/cpp/language/overload_resolution

I think we should avoid pushing too much complexity with everything having just one name into this.

Reply:
Not sure if I understand the problem very well with GetMeter overloads, but can't we add the overloads as done for GetLogger here - https://github.com/open-telemetry/opentelemetry-cpp/blob/a4961c421c9687ea5b750c8df3374cd842b1d5a8/api/include/opentelemetry/logs/logger_provider.h#L39C1-L68C1. What is actually breaking with this ?

Reply:
I found out the root cause of the overload issues.

Fixed with this in the SDK:

```
  /*
    Make sure GetMeter() helpers from the API are seen in overload resolution.
  */
  using opentelemetry::metrics::MeterProvider::GetMeter;
```